File: js/sample-todo/src/TodoList.ts
import { isString } from "@meta-sharp/runtime";
import { Enumerable } from "@meta-sharp/runtime";
import type { Priority } from "./Priority";
import { TodoItem } from "./TodoItem";
export class TodoList {
  constructor(readonly name: string) { }

  private readonly _items: TodoItem[] = [];

  get items(): TodoItem[] {
    return this._items;
  }

  get count(): number {
    return this._items.length;
  }

  get pendingCount(): number {
    return Enumerable.from(this._items).count((i: TodoItem) => !i.completed);
  }

  add(title: string, priority: Priority): void;
  add(item: TodoItem): void;
  add(title: string): void;
  add(...args: unknown[]): void {
    if (args.length === 2 && isString(args[0]) && (args[1] === "low" || args[1] === "medium" || args[1] === "high")) {
      const title = args[0] as string;
      const priority = args[1] as Priority;
      this._items.push(new TodoItem(title, false, priority));
      return;
    }
    if (args.length === 1 && args[0] instanceof TodoItem) {
      const item = args[0] as TodoItem;
      this._items.push(item);
      return;
    }
    if (args.length === 1 && isString(args[0])) {
      const title = args[0] as string;
      this._items.push(new TodoItem(title));
      return;
    }
    throw new Error("No matching overload for add");
  }

  findByTitle(title: string): TodoItem | null {
    return Enumerable.from(this._items).firstOrDefault((i: TodoItem) => i.title === title);
  }

  hasPending(): boolean {
    return Enumerable.from(this._items).any((i: TodoItem) => !i.completed);
  }
}

File: SampleTodo/TodoList.cs
namespace SampleTodo;

public class TodoList(string name)
{
    public string Name { get; } = name;

    private readonly List<TodoItem> _items = [];

    public IReadOnlyList<TodoItem> Items => _items;

    public int Count => _items.Count;

    public int PendingCount => _items.Count(i => !i.Completed);

    public void Add(TodoItem item) => _items.Add(item);

    public void Add(string title) => _items.Add(new TodoItem(title));

    public void Add(string title, Priority priority) => _items.Add(new TodoItem(title, Priority: priority));

    public TodoItem? FindByTitle(string title) => _items.FirstOrDefault(i => i.Title == title);

    public bool HasPending() => _items.Any(i => !i.Completed);
}
